design/tracing-design.md

```diff
@@ -0,0 +1,142 @@
+
+Since the tracing backend is pluggable, Contour will not package any particular backend.
+The user is responsible for deploying and operating the tracing backend of their choice, and configuring Contour to make use of it.
+
+## Detailed Design
+The `tracing` section of the Contour config file will look like:
+
+(WIP, this is the rough idea)
+```yaml
+tracing:
+  # Trace format to use. Valid options are opencensus and zipkin.
+  format: [opencensus, zipkin]
+  config:
+    # If tracing.format == opencensus, then this block is expected
+    # to be populated.
+    opencensus:
+      agent-address: otel-collector.default:55678
+    # If tracing.format == zipkin, then this block is expected
+    # to be populated.
+    zipkin:
+      collector-cluster: <Envoy cluster name>
+      collector-endpoint: /api/v2/spans
+      collector-endpoint-version: ...
+```
+
+In the presence of the `tracing` block above, all HTTP connection managers (HCM) will be configured for tracing.
+
+A Zipkin-enabled HCM will include the following as an example:
+```go
+...
+Tracing: &http.HttpConnectionManager_Tracing{
+    Provider: &envoy_config_trace_v3.Tracing_Http{
+        Name: "envoy.tracers.zipkin",
+        ConfigType: &envoy_config_trace_v3.Tracing_Http_TypedConfig{
+            TypedConfig: protobuf.MustMarshalAny(&envoy_config_trace_v3.ZipkinConfig{
+                CollectorCluster:         "default/zipkin/9411/da39a3ee5e",
+                CollectorEndpoint:        "/api/v2/spans",
+                CollectorEndpointVersion: envoy_config_trace_v3.ZipkinConfig_HTTP_JSON,
+            }),
+        },
+    },
+},
+...
+```
+
+An OpenCensus-enabled HCM will include the following as an example:
+```go
+...
+Tracing: &http.HttpConnectionManager_Tracing{
+    Provider: &envoy_config_trace_v3.Tracing_Http{
+    Name: "envoy.tracers.opencensus",
+    ConfigType: &envoy_config_trace_v3.Tracing_Http_TypedConfig{
+        TypedConfig: protobuf.MustMarshalAny(&envoy_config_trace_v3.OpenCensusConfig{
+            OcagentAddress:         "otel-collector.default:55678",
+            OcagentExporterEnabled: true,
+        }),
+    },
+},
+...
+```
+
+For Zipkin, an Envoy cluster corresponding to the Zipkin collector must exist. There are a few different options for creating this cluster:
+- user explicitly defines the cluster as part of a custom bootstrap file
+- Contour config file takes a Kubernetes service name, and a DAG processor adds a cluster for it to the DAG as a root.
+- Contour config file takes an address, and a DAG processor adds a cluster for it to the DAG as a root.
+
+For OpenCensus, Envoy can only be configured once per lifetime (see [this Envoy docs section][6] for information). Attempts to modify the OpenCensus configuration will cause Envoy to NACK config updates which, because of https://github.com/projectcontour/contour/issues/1176, will cause all future config updates to be ignored until Envoy is restarted. 
```

Comment:
Note: I was wrong about #1176 - at some point the behavior has changed and now the behavior is that no updates will be accepted until the incorrect config is removed, at which time Envoy will recover.